src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,208 +19,240 @@ namespace Castle.MicroKernel
+
+		bool IDictionary.IsReadOnly => isReadOnly;
+
+		public static readonly Arguments Empty = new Arguments { isReadOnly = true };
 
-		object ICollection.SyncRoot
+		public void Add(object key, object value)
 		{
-			get { return arguments.SyncRoot; }
+			EnsureWritable();
+			dictionary.Add(key, value);
 		}
 
-		bool IDictionary.IsFixedSize
+		public void Clear()
 		{
-			get { return arguments.IsFixedSize; }
+			EnsureWritable();
+			dictionary.Clear();
 		}
 
-		bool IDictionary.IsReadOnly
+		public Arguments Clone()
 		{
-			get { return arguments.IsReadOnly; }
+			return new Arguments(dictionary);
 		}
 
-		public void Add(object key, object value)
+		void ICollection.CopyTo(Array array, int index)
 		{
-			arguments.Add(key, value);
+			dictionary.CopyTo(array, index);
 		}
 
-		public void Clear()
+		public bool Contains(object key)
 		{
-			arguments.Clear();
+			return dictionary.Contains(key);
 		}
 
-		public bool Contains(object key)
+		IEnumerator IEnumerable.GetEnumerator()
 		{
-			return arguments.Contains(key);
+			return GetEnumerator();
 		}
 
 		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.GetEnumerator();
+			return dictionary.GetEnumerator();
 		}
 
-		public void Remove(object key)
+		/// <summary>
+		/// Inserts a named argument. If the argument already exists it will be overwritten.
+		/// </summary>
+		public virtual Arguments Insert(string key, object value)
 		{
-			arguments.Remove(key);
+			this[key] = value;
+			return this;
 		}
 
-		public Arguments Clone()
+		/// <summary>
+		/// Inserts a set of named and/or typed arguments. If an argument already exists it will be overwritten.
+		/// </summary>
+		public virtual Arguments Insert(IDictionary values)
 		{
-			var dictionary = arguments as Dictionary<object, object>;
-			if (dictionary != null)
+			foreach (DictionaryEntry entry in values)
 			{
-				var comparerExtended = dictionary.Comparer as ArgumentsComparerExtended;
-				if (comparerExtended != null)
+				if (entry.Key is string || entry.Key is Type)
+				{
+					this[entry.Key] = entry.Value;
+				}
+				else
 				{
-					return new Arguments(arguments, comparerExtended.CustomComparers);
+					throw new ArgumentException($"The argument '{entry.Key}' should be of type string or System.Type");
 				}
 			}
+			return this;
+		}
 
-			return new Arguments(arguments);
+		/// <summary>
+		/// Inserts a set of named arguments, <see cref="Dictionary{TKey,TValue}"/> implements <see cref="IReadOnlyDictionary{TKey,TValue}"/>.
+		/// </summary>
+		public virtual Arguments InsertNamed(IReadOnlyDictionary<string, object> values)
+		{
+			foreach (var entry in values)
+			{
+				this[entry.Key] = entry.Value;
+			}
+			return this;
 		}
 
-		void ICollection.CopyTo(Array array, int index)
+		/// <summary>
+		/// Inserts a set of named arguments, this supports plain types and anonymous types.
+		/// </summary>
+		public virtual Arguments InsertProperties(object instance)
```

Comment:
> The name also sounds very imperative, it's surprising to see this in a fluent API that allows chaining.

I agree, I guess the name Insert was used because there were a few other utility extension methods named `Insert` and that name was followed probably because it wasn't going to be a fluent interface even thought it already was since it return the instance.

The interesting problem with the current API you may be alluding to is that what does this method do without looking at the signature, I know usually the index comes first but is that inserting "B" at index 2?
```
args.Insert("B", 2).
```

If instead we went with `AddNamed` and `AddTyped` for all the utility like methods, and only have the one `Add` method that implements the dictionary interface (maybe a second that adds an `IDictionary` which has both named and typed arguments).

Both options read well then:
```
new Arguments("A", "1")
    .AddNamed("B", 2");

args.AddNamed("B", 2);
```

@jnm2 I'm happy to make this change, do you support it?

Reply:
I do like this, yes!

Reply:
Done